include/swift/Driver/Action.h

```diff
@@ -329,30 +329,44 @@ class GeneratePCHJobAction : public JobAction {
 class DynamicLinkJobAction : public JobAction {
   virtual void anchor();
   LinkKind Kind;
+  file_types::ID inputType;
+  bool LTO;
 
 public:
-  DynamicLinkJobAction(ArrayRef<const Action *> Inputs, LinkKind K)
+  DynamicLinkJobAction(ArrayRef<const Action *> Inputs, LinkKind K, file_types::ID inputType, bool LTO)
```

Comment:
I don’t think that you’ve explained why the input type is needed.  The usage in a few places feels like it would drop files and you already do two passes for the BC and object files.  What’s the intent for this?

Reply:
The driver doesn't do `DynamicLinkJobAction` twice. I'm not clear that what you mean "passes".

The `inputType` is used to drop non-compiler output files.
For example, pass only bitcode files to linker dropping other file types when lto.

Reply:
You extract the `TY_Object` and `TY_bitcode` as separate calls.  Im still trying to understand what exactly prevents the simpler approach.

Reply:
I don't have confidence that I correctly understand your suggestion, but I changed the extraction way. https://github.com/apple/swift/pull/31146/commits/4d78938b36439ea4ed900477581cfaf551954047

If the change doesn't match your thought, could you tell me more detail? 🙇

Reply:
Basically the thought reduces down into the following:

The pipeline for Swift currently looks like this:
```
.swift -> [swiftc] -> .obj ->                                  -> [clang] -> [ld] -> .dll
                           -> [swift-autolink-extract] -> .rsp
```

With LTO, this becomes:
```
                           -> [swift-autolink-extract] -> .rsp 
.swift -> [swiftc] -> .obj ->                                  -> [clang] -> [lld] -> .dll
                           -> [clang] -> [lld] -> .obj
                      .bc
                           -> [swift-autolink-extract] -> .rsp
```

I am asking why is it not enough to do the following:
```
.swift -> [swiftc] -> .objc, .bc -> [clang] -> [lld] -> .obj -> [lld] -> .dll
```

In the last case, simply treating the `.obj` and `.bc` as equivalent types, passing them through the linker driver to the linker.  The linker will emit the object file with the autolink information serialized.  When the linker gets the final set of object files, it will deserialize the arguments and link them appropriately, avoiding the entire `swift-autolink-extract` path and avoiding the special case handling for the bitcode files.  This simplifies the driver, the frontend, and the pipeline.  We do need to know that we are using lld, but that can be accomplished by the use of `-use-ld` indicating that `lld` is being used.  If we treat them identically there is no need for having the input kind information - when you filter the inputs, you want all `TY_Object` and `TY_Bitcode` files.

Reply:
Thanks for detail explanation 🙏 Finally I understood what you want to say.